fuzz/writeout.c

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (C) 2024 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <math.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "writeout.h"
+
+
+static const char* ycbcr_to_str(ASS_YCbCrMatrix ycbcr_val)
+{
+    switch (ycbcr_val) {
+        case YCBCR_NONE:         return "None";
+        case YCBCR_BT601_TV:     return "TV.601";
+        case YCBCR_BT601_PC:     return "PC.601";
+        case YCBCR_BT709_TV:     return "TV.709";
+        case YCBCR_BT709_PC:     return "PC.709";
+        case YCBCR_SMPTE240M_TV: return "TV.240m";
+        case YCBCR_SMPTE240M_PC: return "PC.240m";
+        case YCBCR_FCC_TV:       return "TV.fcc";
+        case YCBCR_FCC_PC:       return "PC.fcc";
+        case YCBCR_DEFAULT:      return "ThereWasNoHeader";
+        case YCBCR_UNKNOWN:
+        default:
+            return "InvalidUnknownValue";
+    }
+}
+
+static bool time_to_str(long time, char** buf, size_t* buf_size)
+{
+    time /= 10; // ASS files can only have centi-second precision
+    int sign = time < 0 ? -1 : 1;
+    time = labs(time);
+
+    // Without chars for hours, except for sign indicator
+    size_t min_size = sign == -1 ? 14 : 10;
+    int cs, s, m;
+    #define DIVMOD(div,mod) mod = time % div; time /= div
+    DIVMOD(100, cs);
+    DIVMOD(60, s);
+    DIVMOD(60, m);
+    #undef DIVMOD
+
+    min_size += time >= 10 ? ceil(log10(time + 1)) : 1;
+    if (*buf_size < min_size) {
+        // +14 are arbitrary padding to avoid too many reallocs
+        char* nbuf = realloc(*buf, min_size + 14);
+        if (!nbuf) {
+            return false;
+        }
+        *buf = nbuf;
+        *buf_size = min_size + 14;
+    }
+
+    int ret = snprintf(*buf, *buf_size, "%ld:%02d:%02d.%02d",
+                       time * sign, m * sign, s * sign, cs * sign);
+    if (ret < 0 || ret >= *buf_size)
+        return false;
+    return true;
+}
+
+static int ssa2ass_align(int ssa_align)
+{
+    return ((ssa_align & 0xC) >> 2) * 3 + (ssa_align & 0x3);
+}
+
+static const char *tracktype_to_str(int track_type)
+{
+    return track_type == TRACK_TYPE_ASS ? "ASS"
+        : track_type == TRACK_TYPE_SSA ? "SSA"
+        : "other";
```

Comment:
Maybe `switch` would be better?

Reply:
yeah, in theory a `switch` could also enable us to get compiler warnings when a new track type value (`ASS2` most likely) gets added, but since this is an inlined anonymous enum and `track_type` thus here only an `int` i don’t think this actually works out.
Alas, changed it to a `switch` and added an explicit distinction between `UNKNOWN` and an out of sync track type list